src/lib.rs

```diff
@@ -274,9 +275,92 @@ fn parse_unary_operator(input: ParseInput) -> ParseResult<UnaryOperator> {
     spanned(op).parse(input)
 }
 
-/// Parses a unary expression. E.g., `-u1:1`, `!u1:1`
-fn parse_unary_expression(input: ParseInput) -> ParseResult<(UnaryOperator, Expression)> {
-    tuple((parse_unary_operator, parse_expression)).parse(input)
+/// Parses unary and atomic expressions. E.g., `-u1:1`, `u1:1`
+fn parse_unary_atomic_expression(input: ParseInput) -> ParseResult<Expression> {
+    // this implementation follows the 'Top Down Operator Precedence' algorithm. See
+    // <https://btmc.substack.com/p/how-to-parse-expressions-easy> or <https://tdop.github.io/>
+    alt((
+        spanned(parse_literal),
+        // TODO parse parenthesized expressions? see
+        // https://btmc.substack.com/p/how-to-parse-expressions-easy
+        spanned(tuple((parse_unary_operator, parse_unary_atomic_expression))),
+    ))
+    .parse(input)
+}
+
+/// Parses a binary operator and the expression that follows it, given the expression preceding
+/// the operator, returning all of this is a binary `Expression`.
+///
+/// E.g. left=`u1:1`, input=`&& u1:1`, returns the Expression for `u1:1 && u1:1`
+///
+/// parse_infix_expression handles any and all kinds of expressions that would
+/// left-recursively call parse_expression
+fn parse_infix_expression(input: ParseInput, left: Expression) -> ParseResult<Expression> {
+    // this implementation follows the 'Top Down Operator Precedence' algorithm. See
+    // <https://btmc.substack.com/p/how-to-parse-expressions-easy> or <https://tdop.github.io/>
+    //
+    // If we had any right-associative operators, we would have to use the 'precedence
+    // lowering' trick found here: https://btmc.substack.com/p/how-to-parse-expressions-easy
+    // Like so: right = parse_expression(tokens, get_precedence(op.tag)-1)
+
+    // TODO I'm going to try and use more combinators/less explicit.
+    // TODO use fn flat_map<G, H, O2>(self, g: G) -> FlatMap<Self, G, O>
+    let (rest, start) = position_ws().parse(input)?;
+    let (rest, op) = parse_binary_operator(rest)?;
+    let (rest, right) = parse_expression(rest, Some(op.thing))?;
+    let (_, end) = nom_locate::position(rest)?;
+    let re = RawExpression::from((left, op, right));
+    Ok((
+        rest,
+        Spanned {
```

Comment:
Would it be possible to create some sort of wrapper that could automatically get start/end and wrap the result in a `Spanned`? Maybe overkill for now (and might hurt readability). Just curious for now.

Reply:
Isn't that exactly what the `spanned` combinator/function does?

As you know after seeing the next PR, this function now uses combinator style, including the `spanned` function.